P4D2_2017_Fall/utils/run_exercise.py

```diff
@@ -0,0 +1,344 @@
+        cli = 'simple_switch_CLI'
+        for sw_name, sw_dict in self.switches.iteritems():
+            if 'cli_input' not in sw_dict: continue
+            # get the port for this particular switch's thrift server
+            sw_obj = self.net.get(sw_name)
+            thrift_port = sw_obj.thrift_port
+
+            cli_input_commands = sw_dict['cli_input']
+            self.logger('Configuring switch %s with file %s' % (sw_name, cli_input_commands))
+            with open(cli_input_commands, 'r') as fin:
+                cli_outfile = '%s/%s_cli_output.log'%(self.log_dir, sw_name)
+                with open(cli_outfile, 'w') as fout:
+                    subprocess.Popen([cli, '--thrift-port', str(thrift_port)],
+                                     stdin=fin, stdout=fout)
+
+    def program_hosts(self):
+        """ Adds static ARP entries and default routes to each mininet host.
+
+            Assumes:
+                - A mininet instance is stored as self.net and self.net.start() has
+                  been called.
+        """
+        for host_name in self.topo.hosts():
+            h = self.net.get(host_name)
+            h_iface = h.intfs.values()[0]
+            link = h_iface.link
+
+            sw_iface = link.intf1 if link.intf1 != h_iface else link.intf2
+            # phony IP to lie to the host about
+            host_id = int(host_name[1:])
+            sw_ip = '10.0.%d.254' % host_id
+
+            # Ensure each host's interface name is unique, or else
+            # mininet cannot shutdown gracefully
+            h.defaultIntf().rename('%s-eth0' % host_name)
+            # static arp entries and default routes
+            h.cmd('arp -i %s -s %s %s' % (h_iface.name, sw_ip, sw_iface.mac))
+            h.cmd('ethtool --offload %s rx off tx off' % h_iface.name)
+            h.cmd('ip route add %s dev %s' % (sw_ip, h_iface.name))
+            h.setDefaultRoute("via %s" % sw_ip)
+
+
+    def do_net_cli(self):
+        """ Starts up the mininet CLI and prints some helpful output.
+
+            Assumes:
+                - A mininet instance is stored as self.net and self.net.start() has
+                  been called.
+        """
+        self.logger("Starting mininet CLI")
+        for h in self.net.hosts:
+            h.describe()
+        # Generate a message that will be printed by the Mininet CLI to make
+        # interacting with the simple switch a little easier.
+        print('')
+        print('======================================================================')
+        print('Welcome to the BMV2 Mininet CLI!')
+        print('======================================================================')
+        print('Your P4 program is installed into the BMV2 software switch')
+        print('and your initial configuration is loaded. You can interact')
+        print('with the network using the mininet CLI below.')
+        print('')
+        print('To inspect or change the switch configuration, connect to')
```

Comment:
I might have to double check, but I think that when running the simple_switch_grpc target (i.e. the target that starts a grpc server for p4 runtime on top of simple_switch) the BMv2 CLI is not usable. In any case, I would change this part of the welcome message when running the simple_switch_grpc target, something like "The switch is running P4 Runtime! To inspect or change the switch configuration, connect to its gRPC server running on <switch IP addr>:<switch grpc port>"

Reply:
confirming that simple_switch_grpc does not start the Thrift server